File: LeetCodeProblems/Problems/LeetCode61RotateList.cs
using LeetCodeProblems.Models;

namespace LeetCodeProblems.Problems;

public class LeetCode61RotateList : BaseProblemClass
{
    public ListNode RotateRight(ListNode head, int k)
    {
        if (head == null || head.next == null)
            return head;
        int counter = 0;
        ListNode current = head;
        ListNode previous = null;
        while (current != null)
        {
            previous=current;
            current = current.next;
            counter++;
        }

        if (k >= counter)
            k %= counter;
        if (k == 0)
            return head;
        var output = GetTail(head, counter-k);
        output.tailPrev.next = null;
        previous.next = head;
        return output.tail;
    }

    private (ListNode tail, ListNode tailPrev) GetTail(ListNode root, int k)
    {
        int counter = 0;
        ListNode current = root;
        ListNode previous = null;
        while (counter < k && current != null)
        {
            previous = current;
            current = current.next;
            counter++;
        }

        return (current, previous);
    }

    public override void Run()
    {
        List<int> input = [1, 2];
        int k = 2;
        var head = ListNode.CreateLinkedLis(input);
        var newHead = RotateRight(head, k);
        ListNode.PrintList(newHead);
    }
}

File: LeetCodeProblemsJava/src/Problems/LeetCode61RotateList.java
package Problems;

public class LeetCode61RotateList extends BaseProblem {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) {
            return head;
        }
        if (head.next == null) {
            return head;
        }
        if (k == 0) {
            return head;
        }
        int length = 0;
        ListNode node = head;
        ListNode tail = null;
        while (node != null) {
            length++;
            tail = node;
            node = node.next;
        }
        int toMove = k % length;
        if (toMove == 0) {
            return head;
        }
        int fromBegin = length - toMove;
        node = head;
        ListNode prev = null;
        while (fromBegin != 0) {
            prev = node;
            node = node.next;
            fromBegin--;
        }
        prev.next = null;
        tail.next = head;
        return node;
    }

    @Override
    public void Run() {
        var head = new ListNode(new int[]{1, 2});
        int k = 5;
        head = rotateRight(head, k);
        Auxileries.Print(head);
    }
}

